### PR TITLE
build(docs-infra): capture all decorators in doc-gen

### LINKED ISSUE #40851: API docs: there's no Optional, Self and SkipSelf decorators
# 📚 Docs or angular.io bug report

### Description
[API docs for @angular/core](https://angular.io/api/core#decorators) don't contain descriptions for Optional, Self and SkipSelf decorators

### What's the affected URL?**
https://angular.io/api/core#decorators

### Expected vs Actual Behavior**
**Actual**: see description section above
**Expected**: there're descriptions for Optional, Self and SkipSelf decorators

## 📷Screenshot
no_self_optional

### PR DESCRIPTION
**Pull request by petebacondarwin**
In 6cff877 we broke the decorator docs because the
doc-gen no longer knew how to identify them.

This commit updates the dgeni processor responsible
for identifying the decorators in the code and ensures
that the docs are now generated correctly.

Fixes #40851

Example of a fixed doc: https://pr41091-618b1d9.ngbuilds.io/api/core/Optional

### CODE REVIEW

File: aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.spec.js [hunk 1]
```diff
@@ -88,4 +87,23 @@ describe('mergeDecoratorDocs processor', () => {
     processor.$process([decoratorDoc, metadataDoc, otherDoc]);
     expect(decoratorDoc.docType).toEqual('decorator');
   });
+
+  it('should handle a type cast before the "make decorator" call', () => {
+    decoratorDoc.symbol.valueDeclaration.initializern = {
```
Comment (gkalpak) [inline aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.spec.js:92]:
```suggestion
    decoratorDoc.symbol.valueDeclaration.initializer = {
```

:grin:

File: aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.spec.js [hunk 2]
```diff
@@ -88,4 +87,23 @@ describe('mergeDecoratorDocs processor', () => {
...
+      expression: decoratorDoc.symbol.valueDeclaration.initializer,
+      type: {},
+    };
+    processor.$process([decoratorDoc, metadataDoc, otherDoc]);
+    expect(decoratorDoc.docType).toEqual('decorator');
+  });
+
+  it('should handle the "make decorator" call being wrapped in a call to `attachInjectFlag()`',
+     () => {
+       decoratorDoc.symbol.valueDeclaration.initializern = {
```
Comment (gkalpak) [inline aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.spec.js:102]:
```suggestion
       decoratorDoc.symbol.valueDeclaration.initializer = {
```

:grin:

File: aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.spec.js [hunk 3]
```diff
@@ -88,4 +87,23 @@ describe('mergeDecoratorDocs processor', () => {
...
+         expression: {test: 'attachInjectFlag'},
+         args: [decoratorDoc.symbol.valueDeclaration.initializer]
```
Comment (gkalpak) [inline aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.spec.js:104]: Shouldn't these be:

```suggestion
         expression: {text: 'attachInjectFlag'},
         arguments: [decoratorDoc.symbol.valueDeclaration.initializer]
```

❓

File: aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.js [hunk 4]
```diff
@@ -109,18 +109,44 @@ module.exports = function mergeDecoratorDocs(log) {
...
+
+  let initializer = declaration.initializer;
+
   // There appear to be two forms of initializer:
-  //    export var Injectable: InjectableFactory =
-  //    <InjectableFactory>makeDecorator(InjectableMetadata);
+  //
+  // ```
+  // export const Injectable: InjectableFactory =
+  //       <InjectableFactory>makeDecorator(InjectableMetadata);
```
Comment (gkalpak) [inline aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.js:123]: That's some arbitrary indentation :grin:

File: aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.js [hunk 5]
```diff
@@ -109,18 +109,44 @@ module.exports = function mergeDecoratorDocs(log) {
...
+
+  // It is also possible that the decorator call is wrapped in a call to `attachInjectFlag()`:
+  //
+  // ```
+  // const Optional: OptionalDecorator =
+  //       attachInjectFlag(makeParamDecorator('Optional'), InternalInjectFlags.Optional);
+  // ```
+  //
+  // If so, use the first argument of the call.
+  if (initializer && initializer.expression && initializer.expression.text === 'attachInjectFlag') {
```
Comment (gkalpak) [inline aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.js:147]: The `initializer` check should be redundant here, right?

```suggestion
  if (initializer.expression && initializer.expression.text === 'attachInjectFlag') {
```

**Review by gkalpak — APPROVED** [review]
A couple of nits and typos. Otherwise lgtm 🚀

File: aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.spec.js (hunk 1 above)
Comment (petebacondarwin) [reply aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.spec.js:92]: 😱

**Comment by petebacondarwin** [comment]
OMG this is the most poor PR I have produced in a long time. The tests were passing because they were not actually checking the right thing, and once that was fixed, it showed a whole load of mistakes! Thanks for the safety net @gkalpak

File: aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.js (hunk 4 above)
Comment (petebacondarwin) [reply aio/tools/transforms/angular-api-package/processors/mergeDecoratorDocs.js:123]: If left to the code formatter it removes all the indent. So I just added a bit to help out.

**Review by gkalpak — APPROVED** [review]
🚀

**Pull request closed without merging**